Add SKE conditions to candidate "accept offer" interface

elliotcm opened the pull request:
This PR should wait until https://github.com/DFE-Digital/apply-for-teacher-training/pull/7944 has been merged, and then rebased onto `main`.

## Changes proposed in this pull request

### Standard SKE subject

<img width="698" alt="Screenshot 2023-02-21 at 16 15 21" src="https://user-images.githubusercontent.com/109225/220411296-e2655c02-dbf8-4037-a0a2-3b9032ff1358.png">
<img width="807" alt="Screenshot 2023-02-21 at 16 15 51" src="https://user-images.githubusercontent.com/109225/220411318-6a7c29b9-a90f-409e-ac17-b183ef2266a7.png">
<img width="1024" alt="Screenshot 2023-02-21 at 16 50 05" src="https://user-images.githubusercontent.com/109225/220411331-6742b278-3ea6-4788-9052-da9f5fd737dc.png">

### Language SKE subject

<img width="856" alt="Screenshot 2023-02-21 at 16 52 32" src="https://user-images.githubusercontent.com/109225/220411385-442e77f3-1c16-4497-a8ba-874baf123ba7.png">
<img width="886" alt="Screenshot 2023-02-21 at 16 56 09" src="https://user-images.githubusercontent.com/109225/220411404-17400669-e518-4cef-9549-5572b30e71ca.png">
<img width="899" alt="Screenshot 2023-02-21 at 16 56 26" src="https://user-images.githubusercontent.com/109225/220411422-7e840745-bf4b-4df0-ac37-03ed0499ecc6.png">

elliotcm commented:
![](https://github.trello.services/images/mini-trello-icon.png) [⛷️ SKE: Candidate - offer journey](https://trello.com/c/1xoU0yQ3/1038-%E2%9B%B7%EF%B8%8F-ske-candidate-offer-journey)

elliotcm commented:
![](https://github.trello.services/images/mini-trello-icon.png) [SKE: Candidate - language offer journey](https://trello.com/c/fVYj5R3i/1203-ske-candidate-language-offer-journey)

elliotcm commented:
To do: write the test for the migration (or delete it and remove all SKE conditions on QA)

JR-G commented on app/models/ske_condition.rb:
```diff
@@ -10,9 +12,52 @@ class SkeCondition < OfferCondition
+    )
+  end
+
+  def course_description
```
might this perhaps be better as a helper?

elliotcm replied:
I suspect this might actually be better as a presenter (including the other formatty methods).

JR-G commented on spec/components/candidate_interface/application_dashboard_course_choices_component_spec.rb:
```diff
@@ -275,6 +275,32 @@
+    let(:course_option) { build(:course_option, course:) }
+    let(:course) { build(:course, course_subjects:) }
+    let(:course_subjects) { [build(:course_subject, subject:)] }
+    let(:subject) { build(:subject, :non_language) } # rubocop:disable RSpec/SubjectDeclaration
```
given our ongoing involvement in the world of education, perhaps this is one to add to the `rubocop.yml`

elliotcm replied:
I'm not sure, that's one for the team.  I think it's a good warning in general, I just couldn't think of another word to use.

JR-G commented on spec/forms/provider_interface/offer_wizard_spec.rb:
```diff
@@ -62,13 +67,15 @@
 
-        it { is_expected.to be_valid(current_step) }
+        it 'be valid' do
+          expect(wizard.valid?(current_step)).to be(true)
```
```suggestion
          expect(wizard).to be_valid(current_step)
```

MaeveRoseveare reviewed (approved):
Looks great, just a few minor things where I couldn't find them in the code to comment on:

1. On the last page for candidates (/manage-conditions), can we remove the 'A/An' in front of the courses
<img width="684" alt="Screenshot 2023-02-23 at 14 52 49" src="https://user-images.githubusercontent.com/68232608/220947704-e9289989-0fa7-41d7-9910-f2b19b394594.png">

2. When I select 1 language course, the text shows up with one bullet, which looks a little odd (like we had on the provider side). Could it be written the same as the normal flow (so no bullets)?
<img width="672" alt="Screenshot 2023-02-23 at 15 08 02" src="https://user-images.githubusercontent.com/68232608/220948089-b444c62d-5988-4acb-8b78-5546fc8fc987.png">

3. Where we have 2 language courses on the 'Details of offer' could we present these with bullets, as its a bit hard to read as one sentence.
<img width="710" alt="Screenshot 2023-02-23 at 15 05 20" src="https://user-images.githubusercontent.com/68232608/220948453-dfcecdf4-7683-472d-9c8d-810244ab82b1.png">

4. All courses besides languages should be lower case (unless at the start of a sentence obvs) e.g. 'physics' - but I see they are capitalised on the provider side too so this might be a separate ticket to change. Not a biggie.

elliotcm commented on app/components/candidate_interface/offer_ske_conditions_review_component.html.erb:
```diff
@@ -0,0 +1,58 @@
+    </p>
+
+    <p class="govuk-body">
+      You should complete the course before your teacher training starts in <%= training_starts.to_fs(:month_and_year) %>.
```
This is the standard flow, the language flow is beneath this.  That said, there are some of those issues there.

elliotcm commented:
@MaeveRoseveare I'll look to do these updates tomorrow.

elliotcm commented:
@MaeveRoseveare I think I've addressed all the points above.
@DFE-Digital/apply-devs Could you re-review?  I made some functional code changes as part of addressing these.

MaeveRoseveare commented:
Looks like my points above have been changed, expect the question about capitalising subjects but I will chat to @frankieroberto abut this when he's back as I think it's bigger than SKE, so leave it for now. :)

I spotted a bug though. When I select 2 language courses on the provider side, when it comes through to the candidate side, it only displays the info for one of the courses twice (in the image below I selected Spanish and German). I also tried with a different set of languages and the same issue happened.

<img width="675" alt="Screenshot 2023-02-27 at 17 15 39" src="https://user-images.githubusercontent.com/68232608/221633922-31207fcd-7326-4d69-bacf-5a2ada50c53c.png">

frankieroberto commented:
@MaeveRoseveare I agree we should avoid capitalising subjects except for languages. However I think it probably requires a bigger technical change, and is complicated by the ITT course titles which often capitalise the subjects and which we don't control... 😬  Suggest we park this issue for now and raise a card in the backlog for it?  

(Similarly, there's a question over whether we should use "maths" over "mathematics"...)
